### PR TITLE
Cranelift

### PR DESCRIPTION
**Pull request by Maaarcocr**
## Why this may be needed

Cranelift is pretty fast JIT code generator created for wasm runtimes that can target x86-64 and arm64 architectures (plus they may support other archs in the future). They also do some minimal amount of optimisations, like getting rid of jumps that go to the next block. 

## How did I add cranelift

I created a new gem `cranelift_ruby` that exposes enough apis from cranelift such that I could implement our compiler backend. The code for the compiler in `Compiler::Cranelift` is incredibly similar to what we do in the x86 one. 

I also had to add some additional information to the bytecode and cfg structures because I needed to know how many transitions we need and also added an index to each push and pop so that I know which variable to write and read from for each transition. 

I'll try to add more comments for the compiler, but I'm opening this to understand if we would like to have this in the first place. It does add one more compiler that we have to maintain (if we do not get rid of the x86 one) and it does add build dependencies (rust and cargo). At the same time, we get to target 2 backends at once (x86 and arm) and get all the benefits from additional optimisations and possibly new architectures without the need to do much on our side.

As of now the cranelift compiler passes all the tests we have and its similar in performance to the x86 compiler.

### CODE REVIEW

**Comment by chrisseaton** [comment]
https://github.com/Maaarcocr/cranelift_ruby

File: bin/parse [hunk 1]
```diff
@@ -5,9 +5,8 @@ $:.unshift(File.expand_path("../lib", __dir__))
 require "optparse"
 require "bundler/setup"
 require "regular_expression"
-require "crabstone"
+# require "crabstone"
```
Comment (chrisseaton) [inline bin/parse:8]: I think we want to add Cranelift and compare against manual generation - rather than remove manual generation.

File: lib/regular_expression/bytecode.rb [hunk 2]
```diff
@@ -38,7 +38,7 @@ def self.compile(nfa)
           builder.mark_label(label[state, index])
 
           if state.transitions.length > 1 && index != state.transitions.length - 1
-            builder.push(Insns::PushIndex.new)
+            builder.push(Insns::PushIndex.new(n_of_transitions))
```
Comment (chrisseaton) [inline lib/regular_expression/bytecode.rb:41]: This makes sense for other backends as well - the x86 backend could use simple `mov` instructions rather than stack operations, using this. It'll reduce false data dependencies.

File: lib/regular_expression/cfg.rb [hunk 3]
```diff
@@ -177,12 +178,12 @@ def to_dot(graph)
...
-            end
-            if (probability = block_exit.metadata[:probability])
-              attributes["penwidth"] = 1 + probability * 4
-            end
+            # if (kind = block_exit.metadata[:kind])
+            #   attributes["color"] = { true_edge: "green", false_edge: "red" }[kind]
+            # end
+            # if (probability = block_exit.metadata[:probability])
+            #   attributes["penwidth"] = 1 + probability * 4
+            # end
```
Comment (chrisseaton) [inline lib/regular_expression/cfg.rb:186]: What was wrong here?

File: lib/regular_expression/compiler/cranelift.rb [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+        def to_proc
+          function = Fiddle::Function.new(@f_ptr, [Fiddle::TYPE_VOIDP, Fiddle::TYPE_SIZE_T], Fiddle::TYPE_SIZE_T)
+          lambda do |string|
+            value = function.call(string, string.length)
+            value if value != string.length + 1
+          end
+        end
+      end
+
+      def self.compile(cfg, _schedule)
```
Comment (chrisseaton) [inline lib/regular_expression/compiler/cranelift.rb:22]: We should make the compiler request a schedule, rather than scheduling always done and then given to the compiler.

**Review by chrisseaton — CHANGES_REQUESTED** [review]

**Comment by chrisseaton** [comment]
Yes would be good to have many backends (since it's a demo project) and see how they compare.

File: bin/parse (hunk 1 above)
Comment (Maaarcocr) [reply bin/parse:8]: yes sorry, I'm not sure why I did this

File: lib/regular_expression/cfg.rb (hunk 3 above)
Comment (Maaarcocr) [reply lib/regular_expression/cfg.rb:186]: It was throwing `hash key "color" is not a Symbol`

**Comment by noahgibbs** [comment]
Huh. You didn't have to mark any benchmark regexps as no-cranelift? If that's correct, that makes it our most robust backend yet -- everything else has failures on the (simple) benchmark regexps except native-Ruby.

**Comment by Maaarcocr** [comment]
> Huh. You didn't have to mark any benchmark regexps as no-cranelift? If that's correct, that makes it our most robust backend yet -- everything else has failures on the (simple) benchmark regexps except native-Ruby.

Actually they were failing but CI was passing, I had to add some no-cranelift benchmark options.

**Comment by noahgibbs** [comment]
Huh. Sounds like I need to make sure a failing benchmark regexp fails CI. I thought it did, but...

File: lib/regular_expression/cfg.rb (hunk 3 above)
Comment (Maaarcocr) [reply lib/regular_expression/cfg.rb]: Maybe it's because I have an old ruby version

Comment (chrisseaton) [reply lib/regular_expression/cfg.rb]: It shouldn't be anything but 3.0.

https://github.com/kddnewton/regular_expression/blob/main/.ruby-version

File: lib/regular_expression/bytecode.rb [hunk 5]
```diff
@@ -14,7 +14,7 @@ def self.compile(nfa)
 
       visited = Set.new
       worklist = [[nfa, [:jump_to_fail]]]
-
+      n_of_transitions = 0
```
Comment (kddnewton) [inline lib/regular_expression/bytecode.rb:17]: Can you move this work into a separate PR? It seems like it would help us in all backends, so we should move this up.

**Comment by Maaarcocr** [comment]
Yesterday I run this on ARM and it worked flawlessly 😄 will have to rebase and I'll add more comments! (Also I'll split off the transition tracking in a smaller PR)

**Comment by kddnewton** [comment]
This looks good to go. Nice work @Maaarcocr this is awesome.
